File: sdks/java/src/main/java/io/coordin8/EventClient.java
package io.coordin8;

import com.google.protobuf.ByteString;
import coordin8.EventOuterClass;
import coordin8.EventServiceGrpc;
import io.grpc.ManagedChannel;
import io.grpc.stub.StreamObserver;

import java.time.Instant;
import java.util.Map;
import java.util.function.Consumer;

public class EventClient {

    private final EventServiceGrpc.EventServiceBlockingStub stub;
    private final EventServiceGrpc.EventServiceStub asyncStub;

    EventClient(ManagedChannel channel) {
        this.stub = EventServiceGrpc.newBlockingStub(channel);
        this.asyncStub = EventServiceGrpc.newStub(channel);
    }

    /**
     * Subscribe to events from a source matching a template.
     *
     * @param source     event source name (e.g. "market.signals")
     * @param template   attribute template to filter events (null = all)
     * @param durable    true for DURABLE (mailbox), false for BEST_EFFORT
     * @param ttlSeconds subscription lease TTL
     * @param handback   opaque data echoed with events (may be null)
     * @return the event registration with lease info
     */
    public EventRegistrationRecord subscribe(String source, Map<String, String> template,
                                             boolean durable, long ttlSeconds, byte[] handback) {
        EventOuterClass.SubscribeRequest.Builder req = EventOuterClass.SubscribeRequest.newBuilder()
                .setSource(source)
                .setDelivery(durable
                        ? EventOuterClass.DeliveryMode.DURABLE
                        : EventOuterClass.DeliveryMode.BEST_EFFORT)
                .setTtlSeconds(ttlSeconds);
        if (template != null) req.putAllTemplate(template);
        if (handback != null) req.setHandback(ByteString.copyFrom(handback));

        EventOuterClass.EventRegistration resp = stub.subscribe(req.build());
        String leaseId = resp.hasLease() ? resp.getLease().getLeaseId() : null;
        return new EventRegistrationRecord(
                resp.getRegistrationId(), resp.getSource(), leaseId, resp.getSeqNum());
    }

    /**
     * Receive events for a registration (server-streaming).
     * The callback is invoked on a gRPC thread for each event.
     *
     * @param registrationId the subscription registration ID
     * @param onEvent        callback for each event
     * @param onError        callback for errors (stream closed)
     */
    public void receive(String registrationId,
                        Consumer<EventRecord> onEvent,
                        Consumer<Throwable> onError) {
        asyncStub.receive(
                EventOuterClass.ReceiveRequest.newBuilder()
                        .setRegistrationId(registrationId)
                        .build(),
                new StreamObserver<EventOuterClass.Event>() {
                    @Override
                    public void onNext(EventOuterClass.Event evt) {
                        Instant emittedAt = evt.hasEmittedAt()
                                ? Instant.ofEpochSecond(evt.getEmittedAt().getSeconds(),
                                                        evt.getEmittedAt().getNanos())
                                : null;
                        onEvent.accept(new EventRecord(
                                evt.getEventId(),
                                evt.getSource(),
                                evt.getEventType(),
                                evt.getSeqNum(),
                                evt.getAttrsMap(),
                                evt.getPayload().toByteArray(),
                                evt.getHandback().toByteArray(),
                                emittedAt));
                    }

                    @Override
                    public void onError(Throwable t) {
                        onError.accept(t);
                    }

                    @Override
                    public void onCompleted() {}
                });
    }

    /**
     * Emit an event into the bus. Any service can call this.
     *
     * @param source    event source name
     * @param eventType event type identifier
     * @param attrs     event attributes (may be null)
     * @param payload   opaque event body (may be null)
     */
    public void emit(String source, String eventType,
                     Map<String, String> attrs, byte[] payload) {
        EventOuterClass.EmitRequest.Builder req = EventOuterClass.EmitRequest.newBuilder()
                .setSource(source)
                .setEventType(eventType);
        if (attrs != null) req.putAllAttrs(attrs);
        if (payload != null) req.setPayload(ByteString.copyFrom(payload));
        stub.emit(req.build());
    }

    /**
     * Renew a subscription lease.
     */
    public LeaseClient.LeaseRecord renewSubscription(String registrationId, long ttlSeconds) {
        coordin8.LeaseOuterClass.Lease lease = stub.renewSubscription(
                EventOuterClass.RenewSubscriptionRequest.newBuilder()
                        .setRegistrationId(registrationId)
                        .setTtlSeconds(ttlSeconds)
                        .build());
        return new LeaseClient.LeaseRecord(
                lease.getLeaseId(),
                lease.getResourceId(),
                lease.hasGrantedAt() ? Instant.ofEpochSecond(lease.getGrantedAt().getSeconds()) : null,
                lease.hasExpiresAt() ? Instant.ofEpochSecond(lease.getExpiresAt().getSeconds()) : null,
                lease.getTtlSeconds());
    }

    /**
     * Cancel a subscription explicitly.
     */
    public void cancelSubscription(String registrationId) {
        stub.cancelSubscription(
                EventOuterClass.CancelSubscriptionRequest.newBuilder()
                        .setRegistrationId(registrationId)
                        .build());
    }

    // ── Records ─────────────────────────────────────────────────────────────

    public record EventRegistrationRecord(
            String registrationId,
            String source,
            String leaseId,
            long seqNum
    ) {}

    public record EventRecord(
            String eventId,
            String source,
            String eventType,
            long seqNum,
            Map<String, String> attrs,
            byte[] payload,
            byte[] handback,
            Instant emittedAt
    ) {}
}

File: sdks/node/src/event-client.ts
import * as grpc from "@grpc/grpc-js";
import { EventServiceClient as GrpcEventClient } from "../gen/coordin8/event";
import type { Event as ProtoEvent } from "../gen/coordin8/event";
import type { Lease } from "./lease-client";

export interface EventRegistration {
  registrationId: string;
  source: string;
  leaseId?: string;
  seqNum: number;
}

export interface EventRecord {
  eventId: string;
  source: string;
  eventType: string;
  seqNum: number;
  attrs: Record<string, string>;
  payload: Buffer;
  handback: Buffer;
  emittedAt?: Date;
}

export interface SubscribeOpts {
  source: string;
  template?: Record<string, string>;
  durable?: boolean;
  ttlSeconds: number;
  handback?: Buffer;
}

function toEvent(evt: ProtoEvent): EventRecord {
  return {
    eventId: evt.eventId,
    source: evt.source,
    eventType: evt.eventType,
    seqNum: evt.seqNum,
    attrs: evt.attrs ?? {},
    payload: evt.payload,
    handback: evt.handback,
    emittedAt: evt.emittedAt ?? undefined,
  };
}

export class EventClient {
  private readonly stub: GrpcEventClient;

  constructor(channel: grpc.Channel) {
    this.stub = new GrpcEventClient(
      "passthrough:///djinn",
      grpc.credentials.createInsecure(),
      { channelOverride: channel }
    );
  }

  /** Subscribe to events from a source matching a template. */
  subscribe(opts: SubscribeOpts): Promise<EventRegistration> {
    return new Promise((resolve, reject) => {
      this.stub.subscribe(
        {
          source: opts.source,
          template: opts.template ?? {},
          delivery: opts.durable ? 0 : 1, // DURABLE=0, BEST_EFFORT=1
          ttlSeconds: opts.ttlSeconds,
          handback: opts.handback ?? Buffer.alloc(0),
        },
        (err, res) => {
          if (err || !res) return reject(err);
          resolve({
            registrationId: res.registrationId,
            source: res.source,
            leaseId: res.lease?.leaseId || undefined,
            seqNum: res.seqNum,
          });
        }
      );
    });
  }

  /**
   * Receive events for a registration (server-streaming).
   * Returns an async iterable of EventRecord objects.
   */
  async *receive(
    registrationId: string,
    signal?: AbortSignal
  ): AsyncGenerator<EventRecord> {
    const stream = this.stub.receive({ registrationId });
    if (signal) {
      signal.addEventListener("abort", () => stream.cancel(), { once: true });
    }
    for await (const raw of stream) {
      yield toEvent(raw);
    }
  }

  /** Emit an event into the bus. */
  emit(
    source: string,
    eventType: string,
    attrs?: Record<string, string>,
    payload?: Buffer
  ): Promise<void> {
    return new Promise((resolve, reject) => {
      this.stub.emit(
        {
          source,
          eventType,
          attrs: attrs ?? {},
          payload: payload ?? Buffer.alloc(0),
        },
        (err) => {
          if (err) return reject(err);
          resolve();
        }
      );
    });
  }

  /** Renew a subscription lease. */
  renewSubscription(
    registrationId: string,
    ttlSeconds: number
  ): Promise<Lease> {
    return new Promise((resolve, reject) => {
      this.stub.renewSubscription(
        { registrationId, ttlSeconds },
        (err, res) => {
          if (err || !res) return reject(err);
          resolve({
            leaseId: res.leaseId,
            resourceId: res.resourceId,
            grantedAt: res.grantedAt ?? undefined,
            expiresAt: res.expiresAt ?? undefined,
            ttlSeconds: res.ttlSeconds,
          });
        }
      );
    });
  }

  /** Cancel a subscription. */
  cancelSubscription(registrationId: string): Promise<void> {
    return new Promise((resolve, reject) => {
      this.stub.cancelSubscription({ registrationId }, (err) => {
        if (err) return reject(err);
        resolve();
      });
    });
  }
}
